cob_base_velocity_smoother/src/cob_base_velocity_smoother/velocity_smoother.cpp

```diff
@@ -46,66 +45,66 @@ VelocitySmoother::VelocitySmoother(const std::string &name)
 
+  robot_feedback = (VelocitySmoother::RobotFeedbackType) config.robot_feedback;
   speed_lim_vx  = config.speed_lim_vx;
   speed_lim_vy  = config.speed_lim_vy;
-  speed_lim_w  = config.speed_lim_w;
+  speed_lim_w   = config.speed_lim_w;
   accel_lim_vx  = config.accel_lim_vx;
   accel_lim_vy  = config.accel_lim_vy;
-  accel_lim_w  = config.accel_lim_w;
-  decel_factor = config.decel_factor;
-  decel_factor_safe = config.decel_factor_safe;
-  decel_lim_vx  = decel_factor*accel_lim_vx;
-  decel_lim_vy  = decel_factor*accel_lim_vy;
-  decel_lim_w  = decel_factor*accel_lim_w;
-  decel_lim_vx_safe = decel_factor_safe*accel_lim_vx;
-  decel_lim_vy_safe = decel_factor_safe*accel_lim_vy;
-  decel_lim_w_safe = decel_factor_safe*accel_lim_w;
+  accel_lim_w   = config.accel_lim_w;
+  decel_factor  = config.decel_factor;
+  safe_factor   = config.safe_factor;
 }
 
 void VelocitySmoother::velocityCB(const geometry_msgs::Twist::ConstPtr& msg)
 {
-  // Estimate commands frequency; we do continuously as it can be very different depending on the
-  // publisher type, and we don't want to impose extra constraints to keep this package flexible
+  boost::mutex::scoped_lock lock(_mutex);
+
+  ros::Time now = ros::Time::now();
+  double duration = (now - last_cb_time).toSec();
+  
   if (period_record.size() < PERIOD_RECORD_SIZE)
   {
-    period_record.push_back((ros::Time::now() - last_cb_time).toSec());
+    period_record.push_back(duration);
   }
   else
   {
-    period_record[pr_next] = (ros::Time::now() - last_cb_time).toSec();
+    period_record[pr_next] = duration;
   }
 
   pr_next++;
   pr_next %= period_record.size();
-  last_cb_time = ros::Time::now();
+  last_cb_time = now;
 
   if (period_record.size() <= PERIOD_RECORD_SIZE/2)
   {
-    // wait until we have some values; make a reasonable assumption (10 Hz) meanwhile
-    cb_avg_time = 0.1;
+    // wait until we have some values; make a reasonable assumption meanwhile
+    cb_avg_time = 1.0/frequency;
   }
   else
   {
     // enough; recalculate with the latest input
     cb_avg_time = median(period_record);
   }
 
-  input_active = true;
-
   // Bound speed with the maximum values
   target_vel.linear.x  =
       msg->linear.x  > 0.0 ? std::min(msg->linear.x,  speed_lim_vx) : std::max(msg->linear.x,  -speed_lim_vx);
   target_vel.linear.y  =
       msg->linear.y  > 0.0 ? std::min(msg->linear.y,  speed_lim_vy) : std::max(msg->linear.y,  -speed_lim_vy);
   target_vel.angular.z =
-      msg->angular.z > 0.0 ? std::min(msg->angular.z, speed_lim_w) : std::max(msg->angular.z, -speed_lim_w);
+      msg->angular.z > 0.0 ? std::min(msg->angular.z, speed_lim_w)  : std::max(msg->angular.z, -speed_lim_w);
+
+  input_active = true;
 }
 
 void VelocitySmoother::odometryCB(const nav_msgs::Odometry::ConstPtr& msg)
 {
+  boost::mutex::scoped_lock lock(_mutex);
```

Comment:
I'm not sure if a mutex protection is needed. Timer and Callbacks are created from the same Nodehandle and the ros spinning model is single threaded. So only one callback at the time can be processed...